### PR TITLE
net: telnet: Switch TELNET shell backend to use sockets 

### PR DESCRIPTION
**Pull request by rlubos**
Rework TELNET shell backend to use socket API for communication and socket service library for socket monitoring.
Improve TELNET command parsing to make it more robust against possible stream fragmentation.

### CODE REVIEW

File: subsys/shell/backends/Kconfig.backends [hunk 1]
```diff
@@ -358,11 +358,22 @@ config SHELL_BACKEND_TELNET
...
+	select NET_SOCKETS
 	help
 	  Enable TELNET backend.
 
 if SHELL_BACKEND_TELNET
 
+config SHELL_TELNET_INIT_PRIORITY
+	int "Initialization priority"
+	default 95
+	range 0 99
```
Comment (jukkar) [inline subsys/shell/backends/Kconfig.backends:371]: Perhaps we could put a range here that takes the service init priority into account, would something like this work (not tested)
```
range NET_SOCKETS_SERVICE_INIT_PRIO 99
```
And as the range does not allow arithmetic, the already existing build assert verifies that the init values are not the same.

Comment (rlubos) [reply subsys/shell/backends/Kconfig.backends]: Ok, it seems that it worked, so updated the range. It's a bit pity we can't do a simple +1 there.

File: subsys/shell/backends/shell_telnet.c [hunk 2]
```diff
@@ -197,22 +191,25 @@ static int telnet_send(void)
...
-		ret = net_context_send(sh_telnet->client_ctx, msg,
-				       len, telnet_sent_cb,
-				       K_FOREVER, NULL);
-
-		if (ret == -EAGAIN || ret == -ENOBUFS) {
-			k_sleep(K_MSEC(TELNET_RETRY_SEND_SLEEP_MS));
-			continue;
+		ret = zsock_send(sh_telnet->fds[SOCK_ID_CLIENT].fd, msg, len,
+				 block ? 0 : ZSOCK_MSG_DONTWAIT);
+		if (!block && (ret < 0) & (errno == EAGAIN)) {
```
Comment (pdgendt) [inline subsys/shell/backends/shell_telnet.c:201]: I think `&&` was intended?

```suggestion
		if (!block && (ret < 0) && (errno == EAGAIN)) {
```

Comment (rlubos) [reply subsys/shell/backends/shell_telnet.c]: True, fixed

**Review by pdgendt — APPROVED** [review]

**Review by jukkar — APPROVED** [review]

**Comment by jukkar** [comment]
ping @jakub-uC, PTAL

**Review by jakub-uC — APPROVED** [review]
I am not a telnet expert :) However it looks good to me.

**Pull request merged**